File: code_reality/transition.py
"""transition diff——兩 snapshot module-edge 集差異＋「EP 宣稱 vs 實際」對照。

UC5 最後一哩＋intent drift 機械化第一步（報告 §6 裁決）：Depwire diff 的
輕量自建版（共享 worktree 危害 R6——本工具只讀 snapshot sidecar，不碰
working tree）。

用法::

    uv run python -m code_reality.transition <a.json> <b.json> \
        [--ep <ep.md>] [--repo PATH] [-o <prefix>]

輸出 markdown（人讀）＋JSON（機讀）雙格式：``<prefix>.md`` / ``<prefix>.json``。
已知未覆蓋：rename 偵測（module 改名視為 remove+add，報告如實標註）。
"""

import argparse
import json
import re
from dataclasses import dataclass, field
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from code_reality.profile import Profile, claims_re, load_profile, module_of

Edge = tuple[str, str, str]

_BASELINE_RE = re.compile(r"\*\*baseline\*\*:\s*([0-9a-f]{7,40})")


@dataclass
class LoadedSnapshot:
    path: Path
    meta: dict[str, Any]
    files: set[str]
    module_edges: set[Edge]


@dataclass
class EdgeDiff:
    added: list[Edge]
    removed: list[Edge]
    reversed: list[tuple[str, str]]  # 一律 added 方向（B1）
    changed_modules: set[str] = field(default_factory=set)


@dataclass
class ClaimsCompare:
    claimed_and_changed: list[str]
    changed_not_claimed: list[str]
    claimed_not_changed: list[str]
    claims_none: bool = False


def load_snapshot(path: Path) -> LoadedSnapshot:
    data = json.loads(path.read_text())
    assert isinstance(data, dict) and "_meta" in data and "module_edges" in data, (
        f"非 S2 snapshot 格式（缺 _meta/module_edges）: {path}"
    )
    edges_raw = data["module_edges"]
    assert all(isinstance(e, list) and len(e) == 3 for e in edges_raw), (
        f"module_edges 元素非 [src, dst, kind] 三元組: {path}"
    )
    return LoadedSnapshot(
        path=path,
        meta=data["_meta"],
        files=set(data.get("files", [])),
        module_edges={tuple(e) for e in edges_raw},
    )


def diff_edges(a: set[Edge], b: set[Edge]) -> EdgeDiff:
    """pair 集合差運算（B1：tuple-diff 投影 ≠ pair 集合差——multi-kind
    重複對時兩者不等價，pair 投影才正確）。"""
    removed = a - b
    added = b - a
    removed_pairs = {(s, d) for s, d, _ in removed}
    added_pairs = {(s, d) for s, d, _ in added}
    reversed_added_dir = sorted(added_pairs & {(d, s) for s, d in removed_pairs})
    changed = {m for pair in removed_pairs | added_pairs for m in pair}
    return EdgeDiff(
        added=sorted(added),
        removed=sorted(removed),
        reversed=reversed_added_dir,
        changed_modules=changed,
    )


def summarize(
    sa: LoadedSnapshot, sb: LoadedSnapshot
) -> tuple[EdgeDiff, list[str], list[str]]:
    """(diff, new_files, gone_files) 計算一次——main 與 render 共用（F5）。"""
    return (
        diff_edges(sa.module_edges, sb.module_edges),
        sorted(sb.files - sa.files),
        sorted(sa.files - sb.files),
    )


def extract_ep_claims(ep_path: Path, profile: Profile | None) -> set[str]:
    """EP markdown 內 ``<prefix>/<mod>`` 路徑 mentions（保守 heuristic：
    抽不到如實 NONE，不腦補）。claims regex 由 profile ``[[module]]``
    prefixes 衍生——無 profile／無規則 → 恆 NONE（generic repo 無前綴
    知識，by design）。"""
    assert ep_path.is_file(), (
        f"EP 檔不存在或非檔案：{ep_path}（SM-12——NONE 是檔在但無 mention）"
    )
    return set(claims_re(profile).findall(ep_path.read_text()))


def extract_baseline(ep_path: Path) -> str | None:
    """EP 檔頭 ``baseline: <hash>`` 欄（execution-plan skill 慣例）。"""
    assert ep_path.is_file(), f"EP 檔不存在或非檔案：{ep_path}"
    m = _BASELINE_RE.search(ep_path.read_text())
    return m.group(1) if m else None


def compare_claims(claims: set[str], changed_modules: set[str]) -> ClaimsCompare:
    return ClaimsCompare(
        claimed_and_changed=sorted(claims & changed_modules),
        changed_not_claimed=sorted(changed_modules - claims),
        claimed_not_changed=sorted(claims - changed_modules),
        claims_none=not claims,
    )


def _changed_modules(
    diff: EdgeDiff, new_files: list[str], gone_files: list[str], profile: Profile | None
) -> set[str]:
    """實際變動模組＝邊拓撲變動 ∪ 檔案增刪所屬模組。

    只看邊拓撲會把「模組加檔案但邊不變」誤報為宣稱未動（intent-drift
    假陰性）——檔案級變動也是實際變動。
    """
    return diff.changed_modules | {
        module_of(f, profile) for f in new_files + gone_files
    }


def _fmt_edges(edges: list[Edge], limit: int = 20) -> list[str]:
    lines = [f"- `{s} -> {d}` ({k})" for s, d, k in edges[:limit]]
    if len(edges) > limit:
        lines.append(f"- ... +{len(edges) - limit} more")
    return lines


def _trunc_lines(entries: list[str], limit: int = 20) -> list[str]:
    """截斷一律附尾行（F6）——靜默截斷讓消費者誤以為列表完整。"""
    lines = [f"- {e}" for e in entries[:limit]]
    if len(entries) > limit:
        lines.append(f"- ... +{len(entries) - limit} more")
    return lines


def render_report(
    sa: LoadedSnapshot,
    sb: LoadedSnapshot,
    claims: set[str] | None,
    diff: EdgeDiff,
    new_files: list[str],
    gone_files: list[str],
    profile: Profile | None = None,
) -> str:
    a8, b8 = sa.meta.get("commit", "?")[:8], sb.meta.get("commit", "?")[:8]
    lines = [
        f"# Transition Report: {sb.meta.get('repo', '?')}",
        "",
        f"- before: `{a8}`（{sa.path.name}）",
        f"- after: `{b8}`（{sb.path.name}）",
        f"- module edges: {len(sa.module_edges)} -> {len(sb.module_edges)}"
        f"（+{len(diff.added)} / -{len(diff.removed)} / reversed {len(diff.reversed)}）",
        f"- files: {len(sa.files)} -> {len(sb.files)}（+{len(new_files)} / -{len(gone_files)}）",
        "",
    ]
    if not (diff.added or diff.removed or diff.reversed or new_files or gone_files):
        lines.append("## 無結構變化")
        lines.append("")
        lines.append("兩 snapshot 邊集與檔案集相同（同 commit 或無結構變動）。")
        lines.append("")
        return "\n".join(lines)

    lines.append("## 邊變化")
    lines.append("")
    if diff.added:
        lines.append(f"### added ({len(diff.added)})")
        lines.extend(_fmt_edges(diff.added))
        lines.append("")
    if diff.removed:
        lines.append(f"### removed ({len(diff.removed)})")
        lines.extend(_fmt_edges(diff.removed))
        lines.append("")
    if diff.reversed:
        lines.append(f"### reversed ({len(diff.reversed)})——added 方向")
        lines.extend(_trunc_lines([f"`{s} <-> {d}`" for s, d in diff.reversed]))
        lines.append("")
    if new_files:
        lines.append(f"### new files ({len(new_files)})")
        lines.extend(_trunc_lines(new_files))
        lines.append("")
    if gone_files:
        lines.append(f"### gone files ({len(gone_files)})")
        lines.extend(_trunc_lines(gone_files))
        lines.append("")
    if diff.removed and diff.added:
        lines.append("> 已知未覆蓋：rename 偵測（module 改名表現為 remove+add）。")
        lines.append("")

    lines.append("## EP 宣稱 vs 實際變動")
    lines.append("")
    if claims is None:
        lines.append("未提供 `--ep`（EP 宣稱模組路徑對照省略）。")
    elif not claims:
        lines.append("claims: **NONE**——EP 內無 profile prefix 路徑 mention。")
        lines.append(
            f"- 實際變動模組（供判讀，無宣稱可比對）：{sorted(_changed_modules(diff, new_files, gone_files, profile))}"
        )
    else:
        cmp = compare_claims(
            claims, _changed_modules(diff, new_files, gone_files, profile)
        )
        lines.append(
            f"- 宣稱命中 ({len(cmp.claimed_and_changed)})：{cmp.claimed_and_changed}"
        )
        lines.append(
            f"- 實際超出——EP 沒提卻變了 ({len(cmp.changed_not_claimed)})：{cmp.changed_not_claimed}"
        )
        lines.append(
            f"- 宣稱未動——EP 說要動但沒變 ({len(cmp.claimed_not_changed)})：{cmp.claimed_not_changed}"
        )
    lines.append("")
    return "\n".join(lines)


def render_json(
    sa: LoadedSnapshot,
    sb: LoadedSnapshot,
    claims: set[str] | None,
    diff: EdgeDiff,
    new_files: list[str],
    gone_files: list[str],
    profile: Profile | None = None,
) -> dict[str, Any]:
    out: dict[str, Any] = {
        "_meta": {
            "tool": "code_reality.transition",
            "created_at": datetime.now(UTC).isoformat(),
            "before": sa.meta.get("commit"),
            "after": sb.meta.get("commit"),
            "repo": sb.meta.get("repo"),
        },
        "added": [list(e) for e in diff.added],
        "removed": [list(e) for e in diff.removed],
        "reversed": [list(e) for e in diff.reversed],
        "changed_modules": sorted(
            _changed_modules(diff, new_files, gone_files, profile)
        ),
        "new_files": new_files,
        "gone_files": gone_files,
    }
    if claims is not None:
        cmp = compare_claims(
            claims, _changed_modules(diff, new_files, gone_files, profile)
        )
        out["ep_claims"] = {
            "claims": sorted(claims),
            "claims_none": cmp.claims_none,
            "claimed_and_changed": cmp.claimed_and_changed,
            "changed_not_claimed": cmp.changed_not_claimed,
            "claimed_not_changed": cmp.claimed_not_changed,
        }
    return out


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("snapshot_a", type=Path, help="before snapshot（S2 schema）")
    parser.add_argument("snapshot_b", type=Path, help="after snapshot（S2 schema）")
    parser.add_argument(
        "--ep", type=Path, default=None, help="EP markdown（宣稱模組對照）"
    )
    parser.add_argument(
        "--repo",
        type=Path,
        default=Path.cwd(),
        help="repo 根（profile 載入——module/claims 規則來源）",
    )
    parser.add_argument(
        "-o",
        "--output-prefix",
        type=Path,
        default=None,
        help="輸出前綴（預設 transition-<a>..<b>）",
    )
    args = parser.parse_args()

    sa, sb = load_snapshot(args.snapshot_a), load_snapshot(args.snapshot_b)
    profile = load_profile(args.repo)
    if args.ep and profile is None:
        print(
            "[WARN] claims 恆 NONE——--repo 未指到含 .code-reality.toml 的 repo，"
            "宣稱對照不生效（--repo 預設 cwd）"
        )
    claims = extract_ep_claims(args.ep, profile) if args.ep else None
    a8, b8 = sa.meta.get("commit", "?")[:8], sb.meta.get("commit", "?")[:8]

    prefix = args.output_prefix or Path(f"transition-{a8}..{b8}")
    prefix.parent.mkdir(parents=True, exist_ok=True)
    md_path = prefix.with_name(prefix.name + ".md")
    json_path = prefix.with_name(prefix.name + ".json")
    diff, new_files, gone_files = summarize(sa, sb)
    md_path.write_text(
        render_report(sa, sb, claims, diff, new_files, gone_files, profile)
    )
    json_path.write_text(
        json.dumps(
            render_json(sa, sb, claims, diff, new_files, gone_files, profile),
            indent=1,
        )
    )
    print(
        f"[OK] transition {a8} -> {b8}: +{len(diff.added)} / -{len(diff.removed)} /"
        f" reversed {len(diff.reversed)} -> {md_path} + {json_path}"
    )
    if args.ep:
        baseline = extract_baseline(args.ep)
        if baseline:
            print(f"[LOG] EP baseline={baseline}（diff before 應錨定此 commit）")
    print(f"[LOG] rg 'changed_not_claimed' {json_path}")


if __name__ == "__main__":
    main()

File: crates/code-reality/src/transition.rs
//! `transition` — the snapshot-diff DOMAIN module (frozen
//! `code_reality/transition.py` semantics, post canonical-sync
//! 281e07e): snapshot pair set-diff (B1 reversed = added direction),
//! EP claims comparison (regex ∪ relative path-token normalization
//! with repo-root existence verification), degenerate/cross-face
//! guards, json render. The CLI/report face was retired (S4,
//! 2026-08-29 user adjudication — no real consumers; the machine
//! comparison rides delta_tour): delta_tour is the sole diff
//! interface, consuming `load_snapshot`/`summarize`/
//! `extract_ep_claims`/`render_json_value`.

use crate::common::utc_now_iso_micros;
use crate::profile::{claims_regex, module_of, Profile};
use serde_json::{json, Map, Value};
use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

pub type Edge = (String, String, String);

#[derive(Debug, Clone)]
pub struct LoadedSnapshot {
    pub path: PathBuf,
    pub meta: Map<String, Value>,
    pub files: BTreeSet<String>,
    pub module_edges: BTreeSet<Edge>,
}

pub fn load_snapshot(path: &Path) -> Result<LoadedSnapshot, String> {
    let text =
        std::fs::read_to_string(path).map_err(|e| format!("{} 讀取失敗：{}", path.display(), e))?;
    let data: Value = serde_json::from_str(&text).map_err(|e| {
        format!(
            "非 S2 snapshot 格式（缺 _meta/module_edges）: {}（{e}）",
            path.display()
        )
    })?;
    let obj = data.as_object().ok_or_else(|| {
        format!(
            "非 S2 snapshot 格式（缺 _meta/module_edges）: {}",
            path.display()
        )
    })?;
    if !obj.contains_key("_meta") || !obj.contains_key("module_edges") {
        return Err(format!(
            "非 S2 snapshot 格式（缺 _meta/module_edges）: {}",
            path.display()
        ));
    }
    let edges_raw = obj["module_edges"].as_array().ok_or_else(|| {
        format!(
            "module_edges 元素非 [src, dst, kind] 三元組: {}",
            path.display()
        )
    })?;
    let mut module_edges = BTreeSet::new();
    for e in edges_raw {
        let arr = e.as_array().ok_or_else(|| {
            format!(
                "module_edges 元素非 [src, dst, kind] 三元組: {}",
                path.display()
            )
        })?;
        if arr.len() != 3 || !arr.iter().all(Value::is_string) {
            return Err(format!(
                "module_edges 元素非 [src, dst, kind] 三元組: {}",
                path.display()
            ));
        }
        module_edges.insert((
            arr[0].as_str().unwrap().to_string(),
            arr[1].as_str().unwrap().to_string(),
            arr[2].as_str().unwrap().to_string(),
        ));
    }
    let files: BTreeSet<String> = obj
        .get("files")
        .and_then(Value::as_array)
        .map(|a| {
            a.iter()
                .filter_map(|v| v.as_str().map(|s| s.to_string()))
                .collect()
        })
        .unwrap_or_default();
    Ok(LoadedSnapshot {
        path: path.to_path_buf(),
        meta: obj["_meta"].as_object().cloned().unwrap_or_default(),
        files,
        module_edges,
    })
}

pub struct EdgeDiff {
    pub added: Vec<Edge>,
    pub removed: Vec<Edge>,
    /// always added-direction (B1)
    pub reversed: Vec<(String, String)>,
    pub changed_modules: BTreeSet<String>,
}

/// Pair set-diff (`transition.py:72-86`): the tuple-diff projection ≠
/// pair set-diff under multi-kind duplicates — the pair projection is
/// the correct reversal test.
pub fn diff_edges(a: &BTreeSet<Edge>, b: &BTreeSet<Edge>) -> EdgeDiff {
    let removed: BTreeSet<&Edge> = a.difference(b).collect();
    let added: BTreeSet<&Edge> = b.difference(a).collect();
    let removed_pairs: BTreeSet<(&str, &str)> = removed
        .iter()
        .map(|(s, d, _)| (s.as_str(), d.as_str()))
        .collect();
    let added_pairs: BTreeSet<(&str, &str)> = added
        .iter()
        .map(|(s, d, _)| (s.as_str(), d.as_str()))
        .collect();
    let mut reversed: Vec<(String, String)> = added_pairs
        .intersection(&removed_pairs.iter().map(|&(s, d)| (d, s)).collect())
        .map(|&(s, d)| (s.to_string(), d.to_string()))
        .collect();
    reversed.sort();
    let mut changed = BTreeSet::new();
    for (s, d) in removed_pairs.iter().chain(added_pairs.iter()) {
        changed.insert(s.to_string());
        changed.insert(d.to_string());
    }
    EdgeDiff {
        added: added.into_iter().cloned().collect(),
        removed: removed.into_iter().cloned().collect(),
        reversed,
        changed_modules: changed,
    }
}

/// Pair summary (S4): the diff faces plus the degenerate marking —
/// marked at summarize time so every consumer (delta_tour today, any
/// future one) inherits the guard without re-deriving it.
pub struct Summary {
    pub diff: EdgeDiff,
    pub new_files: Vec<String>,
    pub gone_files: Vec<String>,
    pub degenerate: Option<String>,
}

pub fn summarize(sa: &LoadedSnapshot, sb: &LoadedSnapshot) -> Summary {
    let diff = diff_edges(&sa.module_edges, &sb.module_edges);
    let new_files: Vec<String> = sb.files.difference(&sa.files).cloned().collect();
    let gone_files: Vec<String> = sa.files.difference(&sb.files).cloned().collect();
    let degenerate = degenerate_pair(sa, sb);
    Summary {
        diff,
        new_files,
        gone_files,
        degenerate,
    }
}

/// Degenerate-pair guard: a snapshot whose participating-file set is empty
/// (the REFERENCES-only collapse) cannot support file-face conclusions.
/// ANY side empty → `Some(warning)` distinguishing which side's list is
/// untrustworthy; the consumer must not draw a "no structural change"
/// conclusion from a both-empty pair (a healthy→degenerate pair still
/// carries its — falsely massive — file lists, marked by the warning
/// instead of erased).
pub fn degenerate_pair(sa: &LoadedSnapshot, sb: &LoadedSnapshot) -> Option<String> {
    match (sa.files.is_empty(), sb.files.is_empty()) {
        (true, true) => {
            Some("兩側 snapshot files 皆空（退化快照）——diff 無意義，勿下「無結構變化」結論".into())
        }
        (true, false) => Some("before 側 snapshot files 空（退化）——gone-files 清單不可信".into()),
        (false, true) => Some("after 側 snapshot files 空（退化）——added-files 清單不可信".into()),
        _ => None,
    }
}

fn files_face(s: &LoadedSnapshot) -> Option<&str> {
    s.meta.get("files_face").and_then(Value::as_str)
}

/// Cross-face guard (S2): the files face is comparable only when both
/// sides carry the same `files_face` meta (absent ≡ the pre-widening
/// structural-only face). module_edges stay comparable either way — the
/// structural kind set never changed.
fn face_mismatch(sa: &LoadedSnapshot, sb: &LoadedSnapshot) -> Option<String> {
    match (files_face(sa), files_face(sb)) {
        (Some(a), Some(b)) if a != b => Some(format!(
            "files 面不同（before={a}／after={b}）——files diff 跨面不可比；module_edges 仍可比（kind 集不變）"
        )),
        (None, Some(_)) | (Some(_), None) => Some(
            "files 面跨版本不可比（一方缺 files_face＝舊 structural-only 面）——files diff 不可信；module_edges 仍可比（kind 集不變）".into(),
        ),
        _ => None,
    }
}

fn crg_generation(s: &LoadedSnapshot) -> Option<&str> {
    s.meta.get("crg_last_updated").and_then(Value::as_str)
}

fn crg_raw_edges(s: &LoadedSnapshot) -> Option<u64> {
    s.meta.get("crg_raw_edges").and_then(Value::as_u64)
}

/// Cross-generation guard (MOS-4 symptom 2): the pair spans two
/// graph.db generations — a rebuild happened between the two snapshot
/// points, and a corpus-shrinking rebuild masquerades as mass phantom
/// deletions in the files diff. Loose semantics: either side missing
/// the `crg_last_updated` fingerprint (pre-fingerprint snapshot)
/// skips the check — old snapshots never false-warn. `last_updated`
/// has a single writer, the graph build path, so a value difference
/// means a real rebuild (head-sync refresh touches no meta).
fn generation_mismatch(sa: &LoadedSnapshot, sb: &LoadedSnapshot) -> Option<String> {
    let (ga, gb) = (crg_generation(sa)?, crg_generation(sb)?);
    if ga == gb {
        return None;
    }
    let edges_note = match (crg_raw_edges(sa), crg_raw_edges(sb)) {
        (Some(na), Some(nb)) => format!("；raw edges {na}→{nb}"),
        _ => String::new(),
    };
    Some(format!(
        "before/after 跨 graph 世代（before {ga}／after {gb}{edges_note}）——graph.db 曾在兩次 snapshot 之間重建，檔案集收縮可能是重建造成而非真實刪檔（phantom 刪檔風險）；delta 僅供參考，建議重建後雙端重 snapshot。不擋執行。"
    ))
}

/// File-path token regex (`transition.py:100`).
fn file_token_re() -> &'static regex::Regex {
    static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| regex::Regex::new(r"[A-Za-z0-9_][\w./+-]*\.[A-Za-z0-9]+").unwrap())
}

/// Relative path tokens → module claims with existence verification
/// (`transition.py:103-132`, sync 281e07e): prefix-direct hits resolve
/// as-is; bare-relative tokens resolve only when
/// `repo_root/<prefix>/<first-segment>` is a real directory — a grounded
/// mapping, no guessing.
pub fn path_token_claims(text: &str, profile: &Profile, repo_root: &Path) -> BTreeSet<String> {
    let mut claims = BTreeSet::new();
    for tok in file_token_re().find_iter(text).map(|m| m.as_str()) {
        if !tok.contains('/') {
            continue; // bare filenames cannot be mapped to a module
        }
        let mut resolved: Option<String> = None;
        for rule in &profile.modules {
            if tok.starts_with(&rule.prefix) {
                resolved = Some(tok.to_string());
                break;
            }
        }
        if resolved.is_none() {
            let seg = tok.split('/').next().unwrap_or("");
            for rule in &profile.modules {
                if repo_root.join(&rule.prefix).join(seg).is_dir() {
                    resolved = Some(format!("{}{}", rule.prefix, tok));
                    break;
                }
            }
        }
        if let Some(r) = resolved {
            let m = module_of(&r, Some(profile));
            if !m.is_empty() {
                claims.insert(m);
            }
        }
    }
    claims
}

/// EP claims (`transition.py:135-150`): regex findall ∪ path tokens when
/// `repo_root` is provided. Missing EP file → crash (SM-12: NONE means
/// the file exists but has no mentions).
pub fn extract_ep_claims(
    ep_path: &Path,
    profile: Option<&Profile>,
    repo_root: Option<&Path>,
) -> Result<BTreeSet<String>, String> {
    if !ep_path.is_file() {
        return Err(format!(
            "EP 檔不存在或非檔案：{}（SM-12——NONE 是檔在但無 mention）",
            ep_path.display()
        ));
    }
    let text = std::fs::read_to_string(ep_path)
        .map_err(|e| format!("{} 讀取失敗：{}", ep_path.display(), e))?;
    let mut claims: BTreeSet<String> = claims_regex(profile)
        .find_iter(&text)
        .map(|m| m.as_str().to_string())
        .collect();
    if let (Some(root), Some(p)) = (repo_root, profile) {
        claims.extend(path_token_claims(&text, p, root));
    }
    Ok(claims)
}

pub struct ClaimsCompare {
    pub claimed_and_changed: Vec<String>,
    pub changed_not_claimed: Vec<String>,
    pub claimed_not_changed: Vec<String>,
    pub claims_none: bool,
}

pub fn compare_claims(claims: &BTreeSet<String>, changed: &BTreeSet<String>) -> ClaimsCompare {
    ClaimsCompare {
        claimed_and_changed: claims.intersection(changed).cloned().collect(),
        changed_not_claimed: changed.difference(claims).cloned().collect(),
        claimed_not_changed: claims.difference(changed).cloned().collect(),
        claims_none: claims.is_empty(),
    }
}

/// Actual changed modules = edge topology ∪ file add/remove owners
/// (`transition.py:169-179`).
fn changed_modules_all(
    diff: &EdgeDiff,
    new_files: &[String],
    gone_files: &[String],
    profile: Option<&Profile>,
) -> BTreeSet<String> {
    let mut out = diff.changed_modules.clone();
    for f in new_files.iter().chain(gone_files) {
        out.insert(module_of(f, profile));
    }
    out
}

/// Render the json diff face from a `Summary` (single derivation
/// source: `degenerate` marked once in `summarize`, consumed here —
/// delta_tour pipes this into `build_tour`).
pub fn render_json_value(
    sa: &LoadedSnapshot,
    sb: &LoadedSnapshot,
    summary: &Summary,
    claims: Option<&BTreeSet<String>>,
    profile: Option<&Profile>,
) -> Value {
    let diff = &summary.diff;
    let new_files = &summary.new_files;
    let gone_files = &summary.gone_files;
    let changed_set = changed_modules_all(diff, new_files, gone_files, profile);
    let changed: Vec<String> = changed_set.iter().cloned().collect();
    let mut out = json!({
        "_meta": {
            "tool": "code_reality.transition",
            "created_at": utc_now_iso_micros(),
            "before": sa.meta.get("commit").cloned().unwrap_or(Value::Null),
            "after": sb.meta.get("commit").cloned().unwrap_or(Value::Null),
            "repo": sb.meta.get("repo").cloned().unwrap_or(Value::Null),
        },
        "added": diff.added.iter().map(|(s, d, k)| json!([s, d, k])).collect::<Vec<_>>(),
        "removed": diff.removed.iter().map(|(s, d, k)| json!([s, d, k])).collect::<Vec<_>>(),
        "reversed": diff.reversed.iter().map(|(s, d)| json!([s, d])).collect::<Vec<_>>(),
        "changed_modules": changed,
        "new_files": new_files,
        "gone_files": gone_files,
    });
    // degenerate-pair guard on the json face: the warning field rides
    // alongside the diff faces — delta_tour's build_tour reads it and
    // injects the warning into the tour description (S4).
    if let Some(w) = &summary.degenerate {
        out.as_object_mut()
            .unwrap()
            .insert("degenerate_warning".into(), json!(w));
    }
    if let Some(w) = face_mismatch(sa, sb) {
        out.as_object_mut()
            .unwrap()
            .insert("files_face_warning".into(), json!(w));
    }
    if let Some(w) = generation_mismatch(sa, sb) {
        out.as_object_mut()
            .unwrap()
            .insert("generation_warning".into(), json!(w));
    }
    if let Some(c) = claims {
        let cmp = compare_claims(c, &changed_set);
        out.as_object_mut().unwrap().insert(
            "ep_claims".into(),
            json!({
                "claims": c.iter().cloned().collect::<Vec<_>>(),
                "claims_none": cmp.claims_none,
                "claimed_and_changed": cmp.claimed_and_changed,
                "changed_not_claimed": cmp.changed_not_claimed,
                "claimed_not_changed": cmp.claimed_not_changed,
            }),
        );
    }
    out
}
